### PR TITLE
Test suggested CI integration

### PR DESCRIPTION
**Pull request by keradus**
even if bash minimal integration command is small, it's tricky and have few edge-cases. instead of checking them all manually with every modification of suggested integration, let us start testing them with automate tests.

---------------

BTW, I tried bats.
Sadly, I don't find it nice.
One think is that repo is almost dead, last release 2 years ago.
Other thing is functionallity. As long as all assertions come fine, it's ok. but when sth fail, you want to know as much details as possible.
This is example output:
```
 ✗ check changed files
   (in test file functional/foo.bats, line 21)
     `[ "$output" = "asd" ]' failed
```

no way to check what was value of `$output`

### CODE REVIEW

File: tests/Fixtures/ci-integration/setUp.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+cd $( cd -P "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
+
+git init -q
+git config user.name test
+git config user.email test
+git add .
+git commit -m "init" -q
+
+git checkout -b case1 -q
```
Comment (keradus) [inline tests/Fixtures/ci-integration/setUp.sh]: by adding more and more branches, it's possible to create next and next test cases

File: tests/CiIntegrationTest.php [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+use Symfony\Component\Process\Process;
+
+/**
+ * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * @internal
+ *
+ * @coversNothing
+ */
+final class CiIntegrationTest extends TestCase
```
Comment (keradus) [inline tests/CiIntegrationTest.php]: probably will need to be executed only on linux

File: dev-tools/ci-integration/step2-extra_args.sh [hunk 3]
```diff
@@ -0,0 +1 @@
+EXTRA_ARGS=('--path-mode=intersection' '--' "${CHANGED_FILES[@]}")
```
Comment (keradus) [inline dev-tools/ci-integration/step2-extra_args.sh]: I have split original first step into 2 steps, so it will be easier to adjust #2925 to testing script

**Comment by keradus** [comment]
cc @julienfalque 
could you take a look please?

File: tests/CiIntegrationTest.php [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+        $this->assertRegExp(
+            '/^('.preg_quote($optionalIncompatibilityWarning, '/').')?'.preg_quote($executionDetails, '/').'$/',
+            trim($result3['stderr'])
+        );
+        $this->assertRegExp(
+            '/^Checked all files in \d+\.\d+ seconds, \d+\.\d+ MB memory used$/',
+            trim($result3['output'])
+        );
+        $this->assertSame(0, $result3['code']);
```
Comment (julienfalque) [inline tests/CiIntegrationTest.php]: This assertion is not needed because the related `static::executeScript()` call will throw an exception if the exit code of the command is not 0.

I would drop the `$crashOnError` parameter from `executeScript()`, make it always pass `$crashOnError = false` to `executeCommand()` and check the exit code of each command *before* checking output and stderr.

File: tests/CiIntegrationTest.php [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    private static function executeCommand($command, $crashOnError)
+    {
+        $process = new Process($command, static::$fixtureDir);
+        $process->run();
+
+        $result = array(
```
Comment (julienfalque) [inline tests/CiIntegrationTest.php]: Why not simply return `$process`?

File: dev-tools/ci-integration/step1-changed_files.sh [hunk 6]
```diff
@@ -0,0 +1 @@
+IFS=$'\n'; CHANGED_FILES=($(git diff --name-only --diff-filter=ACMRTUXB "${COMMIT_RANGE}")); unset IFS
```
Comment (julienfalque) [inline dev-tools/ci-integration/step1-changed_files.sh]: Splitting the script into single-line files looks overkill to me. Wouldn't it be easier to have a single file instead?

File: tests/CiIntegrationTest.php [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                $result['output']
+            ));
+        }
+
+        return $result;
+    }
+
+    private static function executeScript(array $scriptParts, $crashOnError)
+    {
+        file_put_contents(static::$tmpFilePath, implode("\n", array_merge(array('#!/usr/bin/env bash', 'set -e', ''), $scriptParts)));
```
Comment (julienfalque) [inline tests/CiIntegrationTest.php]: What is the purpose of the `set -e` line?

File: tests/Fixtures/ci-integration/setUp.sh [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
```
Comment (keradus) [inline tests/Fixtures/ci-integration/setUp.sh]: this file was created when i was trying to test this whole behaviour in bash only (with `bats` - tests runner), it requires a setup to be dedicated file
with current php-based `executeCommand` method we could probably run setup with that method instead of dedicated file.
That's what you mean?

File: tests/CiIntegrationTest.php (hunk 4 above)
Comment (keradus) [inline tests/CiIntegrationTest.php]: as we indeed always pass `true`, I would indeed drop the param, but still raise condition and rely on exception class details which is exposing command-line error details

good catch

File: tests/CiIntegrationTest.php (hunk 5 above)
Comment (keradus) [inline tests/CiIntegrationTest.php]: method doesn't need to expose all process details

File: tests/Fixtures/ci-integration/setUp.sh (hunk 8 above)
Comment (julienfalque) [inline tests/Fixtures/ci-integration/setUp.sh]: Yep :)

File: tests/CiIntegrationTest.php (hunk 7 above)
Comment (keradus) [inline tests/CiIntegrationTest.php]: to crash the script when any command of it is failing, not when last one will fail
we already do it in other `.sh` scripts we do have

File: dev-tools/ci-integration/step1-changed_files.sh (hunk 6 above)
Comment (keradus) [inline dev-tools/ci-integration/step1-changed_files.sh]: right
artifact from `bats` approach when it requested each testing command to be separated file

File: tests/CiIntegrationTest.php (hunk 5 above)
Comment (julienfalque) [inline tests/CiIntegrationTest.php]: Well this is for internal purposes only: the method is private and its result not meant to be used outside the class, so it's not that important. IMO wrapping data inside an array complicates things as you have to know the structure of the array. Using the instance would be easy with IDE autocompletion. Anyway it's a detail, if you prefer using an array, just leave it as is :)

Comment (keradus) [inline tests/CiIntegrationTest.php]: you got me with strong typehinting.

File: tests/Fixtures/ci-integration/setUp.sh (hunk 8 above)
Comment (keradus) [inline tests/Fixtures/ci-integration/setUp.sh]: done

File: tests/CiIntegrationTest.php (hunk 4 above)
Comment (keradus) [inline tests/CiIntegrationTest.php]: done

File: tests/CiIntegrationTest.php (hunk 5 above)
Comment (keradus) [inline tests/CiIntegrationTest.php]: and then I sit down to this today.
it may happen that we will extract this `bash-code-runner` to separate class. then, tightening interface with external dependency.

File: dev-tools/ci-integration/step1-changed_files.sh (hunk 6 above)
Comment (keradus) [inline dev-tools/ci-integration/step1-changed_files.sh]: done

Comment (keradus) [inline dev-tools/ci-integration/step1-changed_files.sh]: done
